app/mender.go

```diff
@@ -179,105 +177,77 @@ func (m *Mender) loadAuth() menderError {
 		return nil
 	}
 
-	code, err := m.authMgr.AuthToken()
-	if err != nil {
-		return NewFatalError(errors.Wrap(err, "failed to cache authorization code"))
+	inChan := m.authMgr.GetInMessageChan()
+	ackChan := m.authMgr.GetAckMessageChan(authManagerChannelName)
+
+	// request
+	inChan <- &AuthManagerRequest{
+		Sender: authManagerChannelName,
+		Method: MethodGetAuthToken,
 	}
 
-	m.authToken = code
+	// ack
+	resp := <-ackChan
+	if resp.Method == MethodGetAuthToken && resp.Error != nil {
+		return NewTransientError(resp.Error)
+	}
```

Comment:
it will be consumed by `authManagerResponseConsumer`, which is running in a dedicated goroutine.

Reply:
Hmm, ok. I have to admit I did not fully understand this function. Let me just try to explain what I was thinking, and you can be the judge if it makes sense to change it or not:

The client normally does not listen for new tokens. The channel can still be kept open, but it's not actively listening, and writes to the channel when it's full are just discarded. When the client needs a token, and finds that it's not available with `GET_AUTH_TOKEN`, it launches a `FETCH_AUTH_TOKEN` request, and *then* starts listening for new tokens. When it's available, it calls `GET_AUTH_TOKEN` again. This is done synchronously, not in a separate Go routine, since the client has to wait for the token at this point (there should be a timeout like in the current implementation though).

Does that make sense? This is getting complicated, so maybe it's better to discuss this in a hangout.

Reply:
Implemented as you requested